### ISSUE TITLE
OpenCL image gives different result than buffer

### ISSUE THREAD

**Issue by boxerab** [intel/compute-runtime#379]
I have a test program here:

https://github.com/GrokImageCompression/latke

that performs the exact same image processing algorithm (debayer) using either image or buffer.

Using latest version of Neo driver,  with buffer, I get a correct result (3 channel colour image).
With opencl image, I get the wrong result (mono image: all three channels are equal).

Here is the ocl buffer result:

![edinburgh_bayer png0](https://user-images.githubusercontent.com/95393/99843564-c3a16380-2b3f-11eb-90ee-e3942c9ae834.png)

and here is the ocl image result:

![edinburgh_bayer png0](https://user-images.githubusercontent.com/95393/99843535-b6847480-2b3f-11eb-80c2-bfe13b310a93.png)

**Comment by AdamCetnerowski** [comment]
Please provide build / release driver version where it worked and where it is broken

**Comment by boxerab** [comment]
Thanks, how can I query the driver version?

**Comment by AdamCetnerowski** [comment]
Are you using the posted weekly releases? That number is good enough for us.

**Comment by boxerab** [comment]
Yes, I am using the weekly release. I don't remember the exact version when it was working.

By the way, I just tried the exact same code with my RX 480 card, and I get the correct result.

**Comment by AdamCetnerowski** [comment]
Unfortunately we can't find a working build, so this isn't a regression. This will go into our regular debug queue. We will update the issue if we have more information or a fix.

**Comment by boxerab** [comment]
@AdamCetnerowski thanks - to clarify, you can't find a working build of the latke project ? I can help , with more details.

**Comment by AdamCetnerowski** [comment]
No, we got the workload to work. We can't find a driver version where the results would be correct.

**Comment by boxerab** [comment]
ahhh I see. So, you can repro, great.

**Comment by kgibala** [comment]
Hi @boxerab,

We sincerely apologize for the delayed response.

After thorough investigation, we have confirmed that the described behavior is reproducible on the latest stack. Our analysis indicates that the root cause lies within the application source file `debayerImage.cl`, specifically in the sampler configuration:
```
CONSTANT sampler_t sampler = CLK_NORMALIZED_COORDS_TRUE | CLK_FILTER_LINEAR | CLK_ADDRESS_MIRRORED_REPEAT;
...
apron[apron_read_row][apron_read_col] = read_imageui(input_image_p, sampler, posSrc).s0;
```
As defined by the [OpenCL specification for read_image](https://registry.khronos.org/OpenCL/sdk/2.1/docs/man/xhtml/read_imagei2d.html):

For the forms that take a sampler, the read_imagei and read_imageui calls support a nearest filter only. The filter_mode specified in sampler must be set to **`CLK_FILTER_NEAREST`** otherwise the values returned are undefined.

The current implementation uses `CLK_FILTER_LINEAR`, which constitutes **undefined behavior per the OpenCL specification and represents a workload-level spec violation**, not a driver defect. For reference, the buffer variant (`debayerBuffer.cl`) performs reads via raw `__global uchar*` pointer arithmetic - without a sampler or filter - which is why it produces correct output on the same hardware.

As a recommendation, please update the sampler configuration in `debayerImage.cl` to use `CLK_FILTER_NEAREST` in order to comply with the OpenCL specification.

From the GPU driver perspective, no defect has been identified in the NEO driver code. Accordingly, we are closing this ticket as Not a Driver Bug (Spec Violation).

If you encounter any new issues in the future, please do not hesitate to open a new submission.

We appreciate your cooperation and understanding.

**Comment by boxerab** [comment]
@kgibala thanks very much for your reply. Would you consider logging a warning in `CL_PROGRAM_BUILD_LOG` when the user violates the spec rather than silently failing ?

**Comment by wenju-he** [comment]
> [@kgibala](https://github.com/kgibala) thanks very much for your reply. Would you consider logging a warning in `CL_PROGRAM_BUILD_LOG` when the user violates the spec rather than silently failing ?

https://github.com/llvm/llvm-project/pull/204086 will add a warning in clang when linear filter is used in read_image{i|ui}
